### ISSUE TITLE
[bug] conan build ../.. --test does nothing

### ISSUE THREAD

**Issue by cjdb** [conan-io/conan#3451]
To help us debug your issue please explain:

* **Conan version:** 1.7
* **Operating system:** Ubuntu 18.04.1 LTS
* **Generator:** `cmake_paths`

Trying to test using Conan:

```bash
conan install ../..
conan build ../.. --configure
conan build ../.. --build
conan build ../.. --test # does nothing
conan build ../.. --build --test # builds and runs CTest
```

This contradicts the [attributes documentation](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#should-configure-should-build-should-install-should-test), which states:

```bash
$ conan build . --configure  # only run cmake.configure(). Other methods will do nothing
$ conan build . --build      # only run cmake.build(). Other methods will do nothing
$ conan build . --install    # only run cmake.install(). Other methods will do nothing
$ conan build . --test       # only run cmake.test(). Other methods will do nothing
# They can be combined
$ conan build . -c -b # run cmake.configure() + cmake.build(), but not cmake.install() nor cmake.test()
```

`conanfile.py` used:

```py
from conans import ConanFile, CMake, RunEnvironment, tools
import os

class ProjectConan(ConanFile):
	name = "project"
	version = "1.0"
	settings = "os", "compiler", "arch", "build_type"
	generators = "cmake_paths", "virtualrunenv"
	build_requires = "catch2/[>=2.3.0]@bincrafters/stable"
	requires = "boost/[>=1.68]@conan/stable", "gsl_microsoft/[>=1.0.0]@bincrafters/stable", "range-v3/[>=0.3.6]@ericniebler/stable"
	exports_sources = "CMakeLists.txt", ".clang*", "test/*", "src/*", "cmake/*"
	no_copy_source = True

	def build(self):
		cmake = CMake(self)

		cmake.configure(source_dir = self.source_folder)
		cmake.build()

		env_build = RunEnvironment(self)
		with tools.environment_append(env_build.vars):
			cmake.test()

	def package_info(self):
		self.env_info.PATH.append(os.path.join(self.package_folder, "bin"))
```

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
Yes, indeed a bug, confirmed.

The root cause is that the ``cmake.test()`` calls ``cmake.build()``

```python
def test(self, args=None, build_dir=None, target=None):
        if not self._conanfile.should_test:
            return
        if not target:
            target = "RUN_TESTS" if self.is_multi_configuration else "test"
        self.build(...)
```
and ``cmake.build()`` will do nothing if ``cmake.should_build`` is not True.

I think the cleanest solution would be to move the ``cmake.build()`` functionality into ``cmake._build()``, except the ``should_build`` test, and this private function is the one that would be called by others.

Thanks very much for reporting it, it will be fixed in next release, please provide the ``--build`` arg in the meantime!

**Comment by memsharded** [comment]
By the way, the ``cmake.install()`` has the same problem, it has to be addressed in the same way.

**Comment by memsharded** [comment]
Hi @cjdb 

This has been fixed and released in 1.7.2, please update and try. Thanks!
